src/Lean/Widget/TaggedText.lean

```diff
@@ -0,0 +1,127 @@
+/-
+Copyright (c) 2021 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Authors: Wojciech Nawrocki
+-/
+import Lean.Data.Json.FromToJson
+import Lean.Server.Rpc.Basic
+
+namespace Lean.Widget
+open Server
+
+/-- The minimal structure needed to represent "string with interesting (tagged) substrings".
+Much like Lean 3 [`sf`](https://github.com/leanprover-community/mathlib/blob/bfa6bbbce69149792cc009ab7f9bc146181dc051/src/tactic/interactive_expr.lean#L38),
+but with indentation already stringified. -/
+inductive TaggedText (α : Type u) where
+  | text (s : String)
+  /- Invariant: non-empty and never directly nested. `append #[tag _ append #[]]` is okay. -/
+  | append (as : Array (TaggedText α))
+  | tag (t : α) (a : TaggedText α)
+  deriving Inhabited, BEq
+
+namespace TaggedText
+
+partial def fromJson? [FromJson α] (j : Json) : Except String (TaggedText α) :=
+  (do
```

Comment:
Is there a reason `deriving FromJson` doesn't work?

If we're doing a custom serializer, we could also pick a more compact representation like:
```typescript
type TaggedText<A> = TaggedText<A>[] | string | {tag: A, text: TaggedText<A>}
```

Reply:
`FromJson` can't be derived because of the `Array (TaggedText α)`, where I had to introduce a local recursive instance of `FromJson (TaggedText α)`. Currently `deriving FromJson` will write `def instFromJsonFoo` and then call that on recursive fields, but will not introduce a local instance. For this to work for `Array` we'd need to look for `Traversable` or something. Though I think we should have `Traversable` anyway, the introduce-local-instance approach avoids this. I can fix up the `deriving` handler to do it.

As for a more compact representation, I did try it but observed no performance upgrade (we are bottlenecked on the `TaggedText`-producing pretty-printing) so I stuck with the same one that `FromJson` would emit.

Reply:
I fixed the handler in #601 and will rebase on that, removing the manual encoding. I agree that our serialization scheme is quite inefficient (it is JSON...), but it's also hardly the bottleneck. For the sizes of messages where it would become problematic, it is far trumped by the message production time, i.e. things like delaboration and formatting. I think if we were to fix this while sticking to JSON rather than moving to e.g. Cap N' Proto, the only nice solution would be to write another `deriving FromJson` strategy (since the handlers are parametrised now). But the current handler is in dire need of a refactor already, so overall it would be quite a bit of work to fix something which isn't really the problem. Note though that I switched to a "flat" `TaggedText.append : Array (TaggedText) -> TaggedText` to avoid some unnecessary nesting, so there aren't as many nested objects (it was blowing the JS stack in VSCode internals).

Reply:
> Note though that I switched to a "flat" TaggedText.append : Array (TaggedText) -> TaggedText to avoid some unnecessary nesting

Yes, this is great (same for `text` I hope).  It also gets rid of the "ugly" `as` name.